File: Medium/784. Letter Case Permutation/letterCasePermutation.java
class Solution {
    public List<String> letterCasePermutation(String S) {
        List<String> ans = new ArrayList<>();
        dfs(S.toCharArray(), 0, ans);
        return ans;
    }
    private void dfs(char [] S, int d, List<String> ans){
        if(d == S.length){
            ans.add(new String(S));
            return;
        }
        dfs(S, d+1, ans);
        if(!Character.isLetter(S[d])){
            return;
        }
        S[d] ^= (1 << 5);
        dfs(S, d+1, ans);
        S[d] ^= (1 << 5);
    }
}

File: Medium/784. Letter Case Permutation/letterCasePermutation.cpp
/*
參考:https://zxi.mytechroad.com/blog/searching/leetcode-784-letter-case-permutation/
*/
class Solution {
public:
    vector<string> letterCasePermutation(string S) {
        vector<string> ans;
        dfs(S, 0, ans);
        return ans;
    }
private:
    void dfs(string &S, int d, vector<string> &ans){
        if(d == S.size()){
            ans.push_back(S);
            return;
        }
        dfs(S, d+1, ans);
        if(!isalpha(S[d])){
            return;
        }
        S[d] ^= (1 << 5);
        dfs(S, d+1, ans);
        S[d] ^= (1 << 5); //記得要將字母復原
    }
};